### PR TITLE
core: add custom function to check emergency URN

### PR DESCRIPTION
**Pull request by anmartan**
#### Pre-Submission Checklist

- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [x] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:

- [ ] PR should be backported to stable branches
- [x] Tested changes locally
- [ ] Related to issue #XXXX (replace XXXX with an open issue number)

#### Description

Improved URN parsing according to [RFC8141](https://datatracker.ietf.org/doc/html/rfc8141#section-2) that allows an arbitrary number of colons to be present in the URN value. Entire URN NID and NSS parts (except for `urn:` scheme) are stored in the `uri.host` struct field. `rq-components` are not supported.

Example of the parse error when parsing "urn:emergency:service:sos"
```
parse_uri(): bad char ':' in state 3 parsed: urn:emergency:service (21) / urn:emergency:service:sos (25)
```

URN values that have been tested to work:
- `urn:emergency:service:sos`
- `urn:emergency:responder.fire`
- `urn:emergency:media-feature`

### CODE REVIEW

File: src/core/parser/parse_uri.c [hunk 1]
```diff
@@ -237,28 +238,33 @@ int parse_uri(char *buf, int len, struct sip_uri *uri)
...
-		state = URI_HEADERS;   \
-		s = p + 1;             \
-		break;                 \
-	case '&':                  \
-	case '@':                  \
+#define check_host_end                             \
+	case ':':                                      \
+		/* found the host */                       \
+		if(scheme != URN_SCH || found_host == 0) { \
+			uri->host.s = s;                       \
```
Comment (IvanRibakov) [inline src/core/parser/parse_uri.c]: This assigns the rest of the string to host, right?

File: src/core/parser/parse_uri.c [hunk 2]
```diff
@@ -237,28 +238,33 @@ int parse_uri(char *buf, int len, struct sip_uri *uri)
...
+			found_host = 1;                        \
```
Comment (IvanRibakov) [inline src/core/parser/parse_uri.c]: Why is this needed?

Comment (anmartan) [reply src/core/parser/parse_uri.c]: The `scheme != URN` is used so that it behaves as it used to if the schema is not URN.

The `found_host` part is used so the beginning of the host string is only assigned once (when the first ":" symbol is found)

File: src/core/parser/parse_uri.c (hunk 1 above)
Comment (anmartan) [reply src/core/parser/parse_uri.c]: `s`is a char*, so this only marks where the beginning of the string would be.

File: src/core/parser/parse_uri.c [hunk 3]
```diff
@@ -237,28 +237,30 @@ int parse_uri(char *buf, int len, struct sip_uri *uri)
...
-		s = p + 1;             \
-		break;                 \
-	case '&':                  \
-	case '@':                  \
+#define check_host_end             \
+	case ':':                      \
+		/* found the host */       \
+		if(scheme != URN_SCH) {    \
+			uri->host.s = s;       \
+			uri->host.len = p - s; \
```
Comment (IvanRibakov) [inline src/core/parser/parse_uri.c:245]: Just to make sure I understand the final solution - where will the `uri->host` be set in case of URN?

Comment (anmartan) [reply src/core/parser/parse_uri.c:245]: These same values will be assigned at the end of the for loop, as the `state` hasn't changed:

https://github.com/zaleos/kamailio/blob/95dc93a2cedb17d671f5952538b3d39e12544107/src/core/parser/parse_uri.c#L1111-L1112

**Comment by IvanRibakov** [comment]
Looks good to proceed!
